### DISCUSSION TITLE
RLS for n:n relations table

### DISCUSSION THREAD

**Discussion by lsei** [supabase/supabase#3802] [Questions]
This is not especially specific to supabase as it's more of a general RLS question but I'm sure relevant to this community. 

I'm trying to secure a table that contains information about which users belong to which groups but ran into a infinite recusion issue. `infinite recursion detected in policy...`

I want users to be able to query the groups they are in but also see the user_ids of other users in the same group. 

However in order to determine if you are allow to see the row of another user you need to run the rls on a separate row. 

My current (non-working) schema: 

```sql
CREATE TABLE "public"."groups" (
    "id" uuid NOT NULL DEFAULT uuid_generate_v4(),
    "created_at" timestamptz DEFAULT now(),
    "name" text,
    PRIMARY KEY ("id")
);

CREATE TABLE "public"."users_groups" (
    "user_id" uuid NOT NULL,
    "group_id" uuid NOT NULL,
    "joined_at" timestamp DEFAULT now(),
    "status" varchar,
    CONSTRAINT "users_groups_user_id_fkey" FOREIGN KEY ("user_id") REFERENCES "auth"."users"("id"),
    CONSTRAINT "users_groups_group_id_fkey" FOREIGN KEY ("group_id") REFERENCES "public"."groups"("id")
);

ALTER TABLE "public"."groups" ENABLE ROW LEVEL SECURITY;
ALTER TABLE "public"."users_groups" ENABLE ROW LEVEL SECURITY;

CREATE POLICY "Only view your own groups"
ON groups
FOR SELECT USING (
    auth.uid() IN (
        SELECT user_id FROM users_groups
        WHERE group_id = id
    )
);

CREATE POLICY "Only view participants of groups you are part of"
ON users_groups
FOR SELECT USING (
    group_id IN (
    -- Can't query users_groups table here. Recursion 
        SELECT group_id 
        FROM users_groups
        WHERE user_id = auth.uid()
    )
);
```

**Comment by GaryAustin1** [comment]
I don't know if there is a better answer just using RLS, but I ended up solving this by using a rpc function.
I take in "room id" and first do an IF EXISTS (SELECT room_table WHERE uid = auth.uid() AND id = room_id). THEN ....  execute the query to to get all records of users with that room_id in the room table and just return their names (I don't want to return their UIDs).  
MY RLS is uid=auth.uid() for direct selects to the room table.
Oh the function security definer so it skips the RLS.

**Comment by soedirgo** [comment]
Can you try these solutions? https://github.com/supabase/supabase/discussions/3328 https://github.com/supabase/supabase/discussions/1138 In general you'd need some sort of `SECURITY DEFINER` function for self-referencing policies.
